### PR TITLE
Druid Intervals Issues prevents metadata pull

### PR DESCRIPTION
**Pull request by kkalyan**
## Why is the change required ?

Caravel is not refreshing metadata with latest Druid Version.  Because of bug in 0.8 (which is fixed 0.8.2)  this was put in deliberately. Since this is open source project and its more likely that new folks would try with latest druid (which 0.9) this will become default issue. Fix is by introspecting druid version using /status API and setting intervals accordingly. 
## How is it tested ?

Manually patched the code tested with local installation (mac + druid 0.9) and default druid [wikiticker example](http://druid.io/docs/0.9.0/tutorials/quickstart.html). This was not working before the patch. 

Screenshots: 
image

image
## Any dependencies for merging this ?

None
## References:

https://github.com/airbnb/caravel/issues/445
https://github.com/airbnb/caravel/issues/409
https://github.com/airbnb/caravel/issues/481
https://github.com/airbnb/caravel/issues/516

### CODE REVIEW

**Comment by vinjadhav** [comment]
@kkalyan thanks...i tested the patch. i am now able to get the druid datasource metadata.

**Comment by LAlbertalli** [comment]
Reference also #516.

As I suggested in my issue, there should be a configuration flag or the function should detect the druid version to decide how to perform the query.

Assuming that everyone is using the last version of druid is a mistake. Druid is used in production in clusters of several servers. Upgrading such an installation is difficult and requires downtime.

**Comment by kkalyan** [comment]
@LAlbertalli  Thats good point. I'll add the config option via 
http://airbnb.io/caravel/installation.html#configuration,
but I'd prefer to keep the default to work with latest version.

**Comment by LAlbertalli** [comment]
@kkalyan Your approach is sensible but I wonder whether it could break someone's installation.

In general, I agree with your approach that the default behavior should be the correct one, not a work around. But in this specific situation, this could cause some issue. I'd like to see what other thinks.

**Comment by kkalyan** [comment]
Revisiting config option, I think adding config at caravel level is not a good idea, as caravel may be used to work with different versions of Druid. 
There should a field (version no) added at each Druid Cluster, since this more involved change, I'll wait till one of the commiters respond.

**Comment by mistercrunch** [comment]
I agree with the approach of adding a version field to the DruidCluster model and hard coding logic based on it.

**Comment by kkalyan** [comment]
I realized that druid has /status API from which this druid version can be retrieved automatically. 
With this there is no need for for any config.

@mistercrunch please review.

File: caravel/models.py [hunk 1]
```diff
@@ -1017,8 +1021,9 @@ def latest_metadata(self):
         # we need to set this interval to more than 1 day ago to exclude
         # realtime segments, which trigged a bug (fixed in druid 0.8.2).
         # https://groups.google.com/forum/#!topic/druid-user/gVCqqspHqOQ
+        start = (0 if self.cluster.druid_version>='0.8.2' else 1)
```
Comment (LAlbertalli) [inline caravel/models.py]: Take care that 0.10.x is greater than 0.8.2 but not for this comparison

File: caravel/models.py [hunk 2]
```diff
@@ -933,6 +933,10 @@ def get_datasources(self):
         return json.loads(requests.get(endpoint).text)
 
     def refresh_datasources(self):
+        endpoint = (
+            "http://{obj.coordinator_host}:{obj.coordinator_port}/status"
+        ).format(obj=self)
+        self.druid_version = json.loads(requests.get(endpoint).text)['version']
```
Comment (LAlbertalli) [inline caravel/models.py]: I suspect you have to mock this call in the tests otherwise you'll get a failure

**Comment by kkalyan** [comment]
@LAlbertalli thanks for review comments. both issues are addressed now.

**Comment by LAlbertalli** [comment]
@kkalyan : I forgot to suggest to refactor the function and move the check version part in a separate function but you have already done that. Kudos 👍

File: caravel/models.py [hunk 3]
```diff
@@ -1017,6 +1020,11 @@ def get_metric_obj(self, metric_name):
             if m.metric_name == metric_name
         ][0]
 
+    def version_higher(self,v1,v2):
+        v1nums = [int(n) for n in v1.split('.')]
+        v2nums = [int(n) for n in v2.split('.')]
+        return (True if v1nums[0]>=v2nums[0] and v1nums[1]>=v2nums[1] and v1nums[2]>=v2nums[2] else False)
```
Comment (LAlbertalli) [inline caravel/models.py]: This condition is wrong. If you test it you will get:
```
self.version_higher('0.9.0','0.8.2')
False
```

The correct test is:
```
    return v1nums[0]>v2nums[0] or \
        (v1nums[0] == v2nums[0] and v1nums[1]>v2nums[1]) or \
        (v1nums[0] == v2nums[0] and v1nums[1] == v2nums[1] and v1nums[2]>v2nums[2])
```

**Comment by kkalyan** [comment]
Thanks @LAlbertalli - Can you help how to write unit tests ?

**Comment by LAlbertalli** [comment]
I could have a look at it. Today I'm pretty busy though, so could look at it just tomorrow.

**Comment by kkalyan** [comment]
@mistercrunch IMO this PR is ready to get merged.

**Comment by mistercrunch** [comment]
Can you address the code quality items raised by landscape.io?
https://landscape.io/diff/379672

**Comment by mistercrunch** [comment]
Glad we got this in!
img

**Comment by ouoc** [comment]
@kkalyan I've met the same problem. When I refresh druid metadata, the datasource didn't have any metrics.  I read alll this issue, and still not understand to  solve the problem. Could you tell me some details?
Thank you very very much.

**Comment by LAlbertalli** [comment]
@ouoc Which version of Caravel are you using? This patch has been merged in 0.10 so if you are using it (or you have cloned the current master) you should be ok on this side.
Also, which version of Druid? Before 0.8.2 the issue that was preventing metadata pull is still present, so you have to wait one day before successfully pulling the metadata.
Finally, have you configured a MySQL server as the druid metadata storage? I haven't investigated the issue, but when using the Derby format (that is not suggested in production anyway) was causing Caravel to fail due to long response time on druid side. Using MySQL solved the issue (and many other Druid issue). We were on druid 0.8.3 at the time, though. Haven't tested on the 0.9.1 that we are using now.

**Comment by ouoc** [comment]
@LAlbertalli I'm using druid 0.9.1... 
I've changed the druid metadata storage to MySQL, but the problem is still here.

**Pull request merged**